plugins/webp-uploads/picture-element.php

```diff
@@ -0,0 +1,159 @@
+<?php
+/**
+ * Add `picture` element support
+ *
+ * @package webp-uploads
+ *
+ * @since n.e.x.t
+ */
+
+/**
+ * Potentially wrap an image tag in a picture element.
+ *
+ * @since n.e.x.t
+ *
+ * @param string $image         The image tag.
+ * @param string $context       The context for the image tag.
+ * @param int    $attachment_id The attachment id.
+ *
+ * @return string The new image tag.
+ */
+function webp_uploads_wrap_image_in_picture( string $image, string $context, int $attachment_id ): string {
+	if ( 'the_content' !== $context ) {
+		return $image;
+	}
+	$image_meta              = wp_get_attachment_metadata( $attachment_id );
+	$original_file_mime_type = get_post_mime_type( $attachment_id );
+	if ( false === $original_file_mime_type || ! isset( $image_meta['sizes'] ) ) {
+		return $image;
+	}
+
+	// Collect all the sub size image mime types.
+	$mime_type_data = array();
+	foreach ( $image_meta['sizes'] as $size ) {
+		if ( isset( $size['sources'] ) && isset( $size['width'] ) && isset( $size['height'] ) ) {
+			foreach ( $size['sources'] as $mime_type => $data ) {
+				$mime_type_data[ $mime_type ]                         = $mime_type_data[ $mime_type ] ?? array();
+				$mime_type_data[ $mime_type ]['w'][ $size['width'] ]  = $data;
+				$mime_type_data[ $mime_type ]['h'][ $size['height'] ] = $data;
+			}
+		}
+	}
+	$sub_size_mime_types = array_keys( $mime_type_data );
+
+	/**
+	 * Filter the image mime types that can be used for the <picture> element.
+	 *
+	 * Default is: ['image/avif', 'image/webp', 'image/jpeg']. Returning an empty array will skip using the `picture` element.
+	 *
+	 * The mime types will output in the picture element in the order they are provided.
+	 * The original image will be used as the fallback image for browsers that don't support the picture element.
+	 *
+	 * @since n.e.x.t
+	 * @param string[] $mime_types    Mime types than can be used.
+	 * @param int      $attachment_id The id of the image being evaluated.
+	 */
+	$enabled_mime_types = (array) apply_filters(
+		'webp_uploads_picture_element_mime_types',
+		array(
+			'image/avif',
+			'image/webp',
+			'image/jpeg',
+		),
+		$attachment_id
+	);
+
+	$mime_types = array_filter(
+		$enabled_mime_types,
+		static function ( $mime_type ) use ( $sub_size_mime_types ) {
+			return in_array( $mime_type, $sub_size_mime_types, true );
+		}
+	);
+
+	// No eligible mime types.
+	if ( count( $mime_types ) === 0 ) {
+		return $image;
+	}
+
+	// If the original mime types is the only one available, no picture element is needed.
+	if ( 1 === count( $mime_types ) && $mime_types[0] === $original_file_mime_type ) {
+		return $image;
+	}
```

Comment:
For sites that don't support AVIF or WebP, if someone uploads an original in one of those formats, won't all the intermediate sizes be JPEG? If so, do we want this to result in using `picture` or would a simple `img` element be preferred?.

Reply:
I haven't tested with WebP support missing, but my expectation is that you shouldn't be able to upload a WebP image if your server doesn't support it. _This may not be working correctly in the block editor currently_, so for now I am referring to uploading in the media library. I'm not sure I have a test version of PHP that doesn't support WebP.

~98% of WordPress sites are on a host that supports WebP so this is an edge case.

Reply:
@joemcgill - I created this follow up [Issue](https://github.com/WordPress/performance/issues/1258) to work on improving the settings screen in particular when WebP isn't supported.